datadog/dogstatsd/base.py

```diff
@@ -304,59 +304,109 @@ def __init__(
+        # being toggled (by `open_buffer()`, `close_buffer()`, or
+        # `self._disable_buffering` calls).
+        self._buffering_toggle_lock = RLock()
 
         # If buffering is disabled, we bypass the buffer function.
         self._send = self._send_to_buffer
-        if disable_buffering:
-            log.info("Statsd buffering is disabled")
+        self._disable_buffering = disable_buffering
+        if self._disable_buffering:
             self._send = self._send_to_server
+            log.info("Statsd buffering is disabled")
 
         # Start the flush thread if buffering is enabled and the interval is above
         # a reasonable range. This both prevents thrashing and allow us to use "0.0"
         # as a value for disabling the automatic flush timer as well.
-        if not disable_buffering and flush_interval >= MIN_FLUSH_INTERVAL:
-            self._register_flush_thread(flush_interval)
-            log.debug(
-                "Statsd flush thread registered with period of %s",
-                flush_interval,
-            )
-        else:
-            log.info("Statsd periodic buffer flush is disabled")
+        self._flush_interval = flush_interval
+        self._flush_thread_stop = threading.Event()
+        self._start_flush_thread(self._flush_interval)
 
     def disable_telemetry(self):
         self._telemetry = False
 
     def enable_telemetry(self):
         self._telemetry = True
 
-    def _register_flush_thread(self, sleep_duration):
-        def _flush_thread_loop(self, sleep_duration):
-            while True:
-                time.sleep(sleep_duration)
+    # Note: Invocations of this method should be thread-safe
+    def _start_flush_thread(self, flush_interval):
+        if self._disable_buffering or self._flush_interval <= MIN_FLUSH_INTERVAL:
+            log.info("Statsd periodic buffer flush is disabled")
+            return
+
+        def _flush_thread_loop(self, flush_interval):
+            while not self._flush_thread_stop.is_set():
+                time.sleep(flush_interval)
                 self.flush()
 
         self._flush_thread = threading.Thread(
             name="{}_flush_thread".format(self.__class__.__name__),
             target=_flush_thread_loop,
-            args=(self, sleep_duration,),
+            args=(self, flush_interval,),
         )
         self._flush_thread.daemon = True
         self._flush_thread.start()
 
+        log.debug(
+            "Statsd flush thread registered with period of %s",
+            self._flush_interval,
+        )
+
+    # Note: Invocations of this method should be thread-safe
+    def _stop_flush_thread(self):
+        if not self._flush_thread:
+            log.warning("No statsd flush thread to stop")
+            return
+
+        try:
+            self.flush()
+        finally:
+            pass
+
+        self._flush_thread_stop.set()
+
+        self._flush_thread.join()
```

Comment:
Not really a big deal I guess but if two threads are calling `_stop_flush_thread` simultaneously, there is a unlucky chance that `join()` could be called on one the thread while the `_flush_thread` has already be terminated by the others.

Reply:
True that this method isn't thread-safe in isolation but [the comment](https://github.com/DataDog/datadogpy/pull/700/files/5534f60dec123e06273e8b258e8b2869a7b375e4#diff-c9e70dcb34e0c8aa6f036a6d228d459d9b57677a7bb41a78372b42265cc51c55R356) on the method mentions this as a warning and [the only caller is lock-protected](https://github.com/DataDog/datadogpy/pull/700/files/5534f60dec123e06273e8b258e8b2869a7b375e4#diff-c9e70dcb34e0c8aa6f036a6d228d459d9b57677a7bb41a78372b42265cc51c55R393-R404). I would have preferred to use the lock directly in this method like you imply but then the toggle to stop would needed 2 lock acquires (1 for toggle method and 1 for stop method; possibly a reentrant one) which seemed like an overkill and over-complication for a helper only called from a single spot.